File: intellij-lsp/src/com/github/gtache/client/languageserver/serverdefinition/ExeLanguageServerDefinition.scala
package com.github.gtache.client.languageserver.serverdefinition

import com.github.gtache.client.connection.{ProcessStreamConnectionProvider, StreamConnectionProvider}
import com.intellij.openapi.diagnostic.Logger

case class ExeLanguageServerDefinition(ext: String, path: String, args: Array[String]) extends UserConfigurableServerDefinition {
  override def createConnectionProvider(workingDir: String): StreamConnectionProvider = {
    if (streamConnectionProvider == null) {
      streamConnectionProvider = new ProcessStreamConnectionProvider(Seq(path) ++ args, workingDir)
    }
    streamConnectionProvider
  }

  override def toString: String = super.toString + " exe : " + path + " args : " + args.mkString(" ")

  override def toArray: Array[String] = {
    Array(typ, ext, path) ++ args
  }

}

object ExeLanguageServerDefinition extends UserConfigurableServerDefinitionObject {
  private val LOG: Logger = Logger.getInstance(this.getClass)

  override def fromArray(arr: Array[String]): ExeLanguageServerDefinition = {
    if (arr.head == typ) {
      val arrTail = arr.tail
      if (arrTail.length < 2) {
        LOG.warn("Not enough elements to translate into a ServerDefinition : " + arr)
        null
      } else {
        ExeLanguageServerDefinition(arrTail.head, arrTail.tail.head, if (arrTail.length > 2) arrTail.tail.tail else Array())
      }
    } else {
      null
    }
  }

  override def typ: String = "exe"
}

File: intellij-lsp/src/main/kotlin/com/github/gtache/lsp/client/languageserver/serverdefinition/ExeLanguageServerDefinition.kt
package com.github.gtache.lsp.client.languageserver.serverdefinition

import com.github.gtache.lsp.head
import com.github.gtache.lsp.tail
import com.github.gtache.lsp.utils.Utils
import com.intellij.openapi.diagnostic.Logger
import java.util.*

/**
 * Class representing server definitions corresponding to an executable file
 * This class is basically a more convenient way to write a RawCommand
 *
 * @param ext  The extension
 * @param path The path to the exe file
 * @param args The arguments for the exe file
 */
data class ExeLanguageServerDefinition(
    override val ext: String,
    val path: String,
    val args: Array<String>
) : BaseServerDefinition(), CommandServerDefinition {

    companion object : UserConfigurableServerDefinitionObject {
        private val logger: Logger = Logger.getInstance(UserConfigurableServerDefinitionObject::class.java)

        override val typ: String = "exe"

        @JvmStatic
        override val presentableTyp: String = "Executable"

        override fun fromArray(arr: Array<String>): ExeLanguageServerDefinition? {
            if (arr.head == typ) {
                val arrTail = arr.tail
                return if (arrTail.size < 2) {
                    logger.warn("Not enough elements to translate into a ServerDefinition : " + arr.joinToString(" ; "))
                    null
                } else {
                    //TODO for cquery, remove
                    ExeLanguageServerDefinition(
                        arrTail.head,
                        arrTail.tail.head,
                        if (arrTail.size > 2) Utils.parseArgs(arrTail.drop(2)) else emptyArray()
                    )
                }
            } else {
                return null
            }
        }
    }

    override val command: Array<String>

    init {
        val tmp = mutableListOf(path)
        tmp.addAll(args)
        command = tmp.toTypedArray()
    }

    override fun toArray(): Array<String> {
        val tmp = mutableListOf(typ, ext, path)
        tmp.addAll(args)
        return tmp.toTypedArray()
    }

    override fun toString(): String = typ + " : path " + path + " args : " + args.joinToString(" ")


    override fun equals(other: Any?): Boolean = other is ExeLanguageServerDefinition && ext == other.ext && path == other.path &&
            args.contentEquals(other.args)

    override fun hashCode(): Int = Objects.hash(ext, path, args)

}